src/proofs/range_proof_wip.rs

```diff
@@ -0,0 +1,571 @@
+
+@license GPL-3.0+ <https://github.com/KZen-networks/bulletproof/blob/master/LICENSE>
+*/
+
+// based on the paper: https://eprint.iacr.org/2020/735.pdf
+//
+// Bulletproofs (https://eprint.iacr.org/2017/1066) uses the inner product argument.
+// Bulletproofs+ (https://eprint.iacr.org/2020/735.pdf) uses the weighted inner product argument
+// which reduces the overall prover communication by ~15%
+// 
+use curv::arithmetic::traits::{Converter, Modulo};
+use curv::cryptographic_primitives::hashing::hash_sha256::HSha256;
+use curv::cryptographic_primitives::hashing::traits::*;
+use curv::elliptic::curves::traits::*;
+use curv::BigInt;
+use curv::{FE, GE};
+use itertools::iterate;
+use proofs::weighted_inner_product::WeightedInnerProdArg;
+use std::ops::{Shl, Shr};
+use Errors::{self, RangeProofError};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RangeProofWIP {
+    A: GE,
+    weighted_inner_product_proof: WeightedInnerProdArg,
+}
+
+impl RangeProofWIP {
+    pub fn prove(
+        g_vec: &[GE],
+        h_vec: &[GE],
+        G: &GE,
+        H: &GE,
+        mut secret: Vec<FE>,
+        blinding: &[FE],
+        bit_length: usize,
+    ) -> RangeProofWIP {
+
+        let num_of_proofs = secret.len();
+        //num of proofs times bit length
+        let nm = num_of_proofs * bit_length;
+        
+        let g_vec = g_vec.to_vec();
+        let h_vec = h_vec.to_vec();
+
+        let two = BigInt::from(2);
+        let one = BigInt::from(1);
+        let order = FE::q();
+
+        //concat all secrets:
+        secret.reverse();
+        let secret_agg = secret.iter().fold(BigInt::zero(), |acc, x| {
+            acc.shl(bit_length) + x.to_big_int()
+        });
+
+        let aL = (0..nm)
+            .map(|i| {
+                let shr_secret = secret_agg.clone().shr(i);
+                shr_secret.modulus(&two)
+            })
+            .collect::<Vec<BigInt>>();
+        let aR = (0..nm)
+            .map(|i| BigInt::mod_sub(&aL[i], &one, &order))
+            .collect::<Vec<BigInt>>();
+
+        let secret_bits = (0..nm)
+            .map(|i| {
+                let bignum_bit: BigInt = aL[i].clone() & BigInt::one();
+                let byte = BigInt::to_vec(&bignum_bit);
+                byte[0] == 1
+            })
+            .collect::<Vec<bool>>();
+
+        // let mut index: usize = 0;
+        let alpha: FE = ECScalar::new_random();
+        let mut A = H * &alpha;
+        A = g_vec.iter().zip(secret_bits.clone()).fold(A, |acc, x| {
+            if x.1 {
+                acc.add_point(&x.0.get_element())
+            } else {
+                acc
+            }
+        });
+        A = h_vec.iter().zip(secret_bits.clone()).fold(A, |acc, x| {
+            if !x.1 {
+                acc.sub_point(&x.0.get_element())
+            } else {
+                acc
+            }
+        });
+
+        let y = HSha256::create_hash_from_ge(&[&A]);
+        let y_bn = y.to_big_int();
```

Comment:
I don't think you should go from fe-> bn -> fe . 
line 105 should be equal to line 107

Reply:
Done. Removed all such redundant computations.